internal/cache/lru/lrucache_test.go

```diff
@@ -0,0 +1,239 @@
+
+	// Insert another.
+	t.insertAndAssert("queso", testData{Value: 34, DataSize: 5}, 1, []int64{26}, nil)
+
+	// See what's left.
+	ExpectEq(nil, t.cache.LookUp("taco"))
+	ExpectEq(23, t.cache.LookUp("burrito").(testData).Value)
+	ExpectEq(28, t.cache.LookUp("enchilada").(testData).Value)
+	ExpectEq(34, t.cache.LookUp("queso").(testData).Value)
+}
+
+func (t *CacheTest) Overwrite() {
+	t.insertAndAssert("burrito", testData{Value: 23, DataSize: 4}, 0, []int64{}, nil)
+	t.insertAndAssert("taco", testData{Value: 26, DataSize: 20}, 0, []int64{}, nil)
+	t.insertAndAssert("enchilada", testData{Value: 28, DataSize: 20}, 0, []int64{}, nil)
+	t.insertAndAssert("burrito", testData{Value: 33, DataSize: 6}, 0, []int64{}, nil)
+
+	// Increase the DataSize while modifying, so eviction should happen
+	t.insertAndAssert("burrito", testData{Value: 33, DataSize: 12}, 1, []int64{26}, nil)
+
+	ExpectEq(nil, t.cache.LookUp("taco"))
+	ExpectEq(33, t.cache.LookUp("burrito").(testData).Value)
+	ExpectEq(28, t.cache.LookUp("enchilada").(testData).Value)
+}
+
+func (t *CacheTest) TestMultipleEviction() {
+	t.insertAndAssert("burrito", testData{Value: 23, DataSize: 4}, 0, []int64{}, nil)
+	t.insertAndAssert("taco", testData{Value: 26, DataSize: 20}, 0, []int64{}, nil)
+	t.insertAndAssert("enchilada", testData{Value: 28, DataSize: 20}, 0, []int64{}, nil)
+
+	// Increase the DataSize while modifying, so eviction should happen
+	t.insertAndAssert("large_data", testData{Value: 33, DataSize: 45}, 3, []int64{23, 26, 28}, nil)
+
+	ExpectEq(nil, t.cache.LookUp("taco"))
+	ExpectEq(nil, t.cache.LookUp("burrito"))
+	ExpectEq(nil, t.cache.LookUp("enchilada"))
+	ExpectEq(33, t.cache.LookUp("large_data").(testData).Value)
+}
+
+func (t *CacheTest) TestWhenEntrySizeMoreThanCacheMaxSize() {
+	t.insertAndAssert("burrito", testData{Value: 23, DataSize: 4}, 0, []int64{}, nil)
+
+	// Insert entry with size greater than maxSize of cache.
+	t.insertAndAssert("taco", testData{Value: 26, DataSize: MaxSize + 1}, 0, []int64{}, errors.New(lru.InvalidEntrySizeErrorMsg))
+
+	ExpectEq(23, t.cache.LookUp("burrito").(testData).Value)
+}
+
+func (t *CacheTest) TestEraseWhenKeyPresent() {
+	t.insertAndAssert("burrito", testData{Value: 23, DataSize: 4}, 0, []int64{}, nil)
+
+	deletedEntry := t.cache.Erase("burrito")
+
+	ExpectEq(23, deletedEntry.(testData).Value)
+	ExpectEq(nil, t.cache.LookUp("burrito"))
+}
+
+func (t *CacheTest) TestEraseWhenKeyNotPresent() {
+	t.insertAndAssert("burrito", testData{Value: 23, DataSize: 4}, 0, []int64{}, nil)
+
+	deletedEntry := t.cache.Erase("taco")
+	ExpectEq(nil, deletedEntry)
+
+	ExpectEq(23, t.cache.LookUp("burrito").(testData).Value)
+}
+
+// This will detect race if we run the test with `-race` flag.
+// We get the race condition failure if we remove lock from Insert or Erase method.
+func (t *CacheTest) TestRaceCondition() {
+	var wg sync.WaitGroup
+	wg.Add(3)
+
+	go func() {
+		defer wg.Done()
+		for i := 0; i < MaxSize; i++ {
```

Comment:
Is there any reason why we running till MaxSize ? I believe we just want to run for multiple iterations, is that's true then we should use different name like numIterations, count.